Add ability to run docker image in AWS Sagemaker Serverless environment

matthewmayo opened the pull request:
Support AWS Sagemaker Serverless endpoints by adding `RUN chmod o+rwX` per AWS docs https://docs.aws.amazon.com/sagemaker/latest/dg/serverless-endpoints-troubleshooting.html.

## What changes are proposed in this pull request?

Adding `RUN chmod o+rwX` per AWS docs to resolve file permission errors with Sagemaker Serverless.  This is to address:  https://github.com/mlflow/mlflow/issues/5544

## How is this patch tested?

- Updated the Dockerfile locally.
- Ran `docker build ...`
- Pushed image to AWS ECR
- Created Sagemaker Model resource using this image
- Created Sagemaker Serverless Endpoint Config using the Model
- Created Sagemaker Endpoint using the Endpoint Config
- Ensured Serverless endpoint was created and could be invoked
- Created Sagemaker Real-Time (EC2) Endpoint Config using the Model
- Created Sagemaker Real-Time Endpoint using the Endpoint Config
- Ensured Real-Time endpoint was created and could be invoked

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

matthewmayo committed:
Update docker_utils.py

Support AWS Sagemaker Serverless endpoints by adding `RUN chmod o+rwX` per AWS docs https://docs.aws.amazon.com/sagemaker/latest/dg/serverless-endpoints-troubleshooting.html.

BenWilson2 commented on mlflow/models/docker_utils.py:
```diff
@@ -40,6 +40,9 @@
 {install_mlflow}
 
 {custom_setup_steps}
+
```
Can you add a note here explaining why this is needed for future viewers of this command?
i.e. 
`# granting read/write access and conditional execution authority to all child directories and files to allow for deployment to <insert name of AWS serverless management service here with a link to the service's main page>` or something along those lines.

matthewmayo committed:
add_comment

Add a comment to Dockerfile template text to add context to why this change is required.

matthewmayo committed:
adding_sign_off

Formatting comment changes, and adding sign off as required...

Signed-off-by: matthewmayo <mmayo1986@gmail.com>

matthewmayo committed:
revert deletion of mleap.py

Accidental deletion of mleap.py restored.  Only change required is in docker_utils.py file.

Signed-off-by: matthewmayo <mmayo1986@gmail.com>

pjroth commented:
@matthewmayo What do you think about updating the title to be something along the lines of the following to better describe the behavior change this PR will make when merged?
> Add ability to run docker images in AWS Sagemaker Serverless environment

BenWilson2 reviewed (approved):
LGTM!
Thank you for the addition!
